java-checks/src/main/java/org/sonar/java/checks/StaticMethodCheck.java

```diff
@@ -95,6 +103,37 @@ public void visitMethod(MethodTree tree) {
     }
   }
 
+  /**
+   * Method's return type requires type parameter and therefore cannot be made static.
+   */
+  private static boolean returnRequiresParentTypeParameter(Symbol.MethodSymbol symbol) {
+    // called from visitMethod, so we have declaration (NPE not possible)
+    ClassTree parent = (ClassTree) symbol.declaration().parent();
+    Set<Symbol> parentTypeParam =
+      parent
+        .typeParameters()
+        .stream()
+        .map(TypeParameterTree::symbol)
+        .collect(Collectors.toUnmodifiableSet());
+    Symbol.TypeSymbol returnType = symbol.returnType();
+    List<Type> returnTypeVars = new ArrayList<>();
+    collectTypeVars(returnTypeVars, returnType.type());
```

Comment:
Open question: Would it be more efficient to collect the type variables in the return type before moving back up the tree and recovering the parent type parameters? I am not sure I can see a point in the flow of the check that would prevent us from entering this block if the method has no type parameter. In such a case, the collection of type parameters from the parent class would be useless here
```suggestion
    List<Type> returnTypeVars = new ArrayList<>();
    collectTypeVars(returnTypeVars, returnType.type());
    if (returnTypeVars.isEmpty()) {
      return false;
    }
    ClassTree parent = (ClassTree) symbol.declaration().parent();
    Set<Symbol> parentTypeParam =
      parent
        .typeParameters()
        .stream()
        .map(TypeParameterTree::symbol)
        .collect(Collectors.toUnmodifiableSet());
    Symbol.TypeSymbol returnType = symbol.returnType();
```